### PR TITLE
[BEAM-428] InProcessRunner - Bundle based local runner

### PR DESCRIPTION
**Pull request by aaltay**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
InProcessRunner is a bundle based runner for doing local execution. It will eventually be a drop-in replacement for the DirectRunner.

It executes pipelines by dividing inputs into bundles and running ptransforms in bundles. Similar to its Java counterpart it aims to provide better performance and validation capabilities compared to the current DirectRunner. InProcessRunner is self contained in its own directory. This will be the case until the replacement happens.

Tests mainly consists of wrapping existing DirectRunner tests and executing them against this new runner.

This version supports batch execution only, future updates will add streaming supports.

### CODE REVIEW

**Comment by aaltay** [comment]
R: @silviulica could you please review?

File: sdks/python/apache_beam/io/fileio.py [hunk 1]
```diff
@@ -521,6 +523,10 @@ def _rename_file(rename_op):
         return(None, e)
       return (final_name, None)
 
+    # ThreadPool crashes in old versions of Python (< 2.7.5) if created from a
```
Comment (silviulica) [inline sdks/python/apache_beam/io/fileio.py]: This seems like a different thing. Isn't it? Better do it in a separate CL unless it blocks the inprocess runner.

**Comment by silviulica** [comment]
LGTM Thanks Ahmet! This adds the inprocess runner in runners/inprocess/* and does not hook it to anything else.

**Comment by aaltay** [comment]
Thank you for the review.

R: @dhalperi Please take a look.

**Comment by dhalperi** [comment]
Merged; please manually close PR.

**Comment by aaltay** [comment]
Thank you!
